### PR TITLE
US69204 pinning API call

### PR DESCRIPTION
**Pull request by ryantmer**
In order for the course tile to do the API call in a cleaner way, I reworked the tile so that the widget just passes in the entire enrollment Entity. Changing the enrollment (i.e. setting it) triggers the observer to parse the enrollment and update the pin button as required. Clicking the pin/unpin button triggers the API request, and updates the pin button as required. The API reply also updates the pin button as required.

Disclaimer: haven't tested this in the LMS yet. Once I overcome my SQL woes, I'll also take a look at amping up the demo so that all of these features will work in that by mocking the token/API requests.

### CODE REVIEW

File: d2l-all-courses.html [hunk 1]
```diff
@@ -1,9 +1,12 @@
...
+<link rel="import" href="../d2l-siren-parser/d2l-siren-parser.html">
+<link rel="import" href="../d2l-ajax/d2l-ajax.html">
+<link rel="import" href="../vui-typography/typography.html">
 <link rel="import" href="d2l-all-courses-styles.html">
 <link rel="import" href="d2l-simple-overlay.html">
 <link rel="import" href="d2l-course-tile.html">
-<link rel="import" href="../d2l-ajax/d2l-ajax.html">
-<link rel="import" href="../vui-typography/typography.html">
+<link rel="import" href="d2l-formatjs-behavior.html">
+<link rel="import" href="d2l-lang-terms.html">
```
Comment (d2l-joelp) [inline d2l-all-courses.html]: We are not using `d2l-formatjs-behavior` or `d2l-lang-terms` anymore (as of this morning). Did you rebase?

File: d2l-all-courses.html [hunk 2]
```diff
@@ -98,14 +91,13 @@ <h3 class="header">{{localize('unpinnedCourses')}}</h3>
...
 					this._setUnpinnedCoursesEntities(
 						allCoursesEntities.filter(
 							function(element) {
-								var isUnpinned = this.getPinnedState(element.entities) !== true;
-								return isUnpinned;
-							}.bind(this)
-						)
-					);
+								var entity = parser.parse(element);
+								return !entity.hasEntityByClass('pinned');
```
Comment (d2l-joelp) [inline d2l-all-courses.html]: So clean!

File: d2l-course-tile.html [hunk 3]
```diff
@@ -1,86 +1,143 @@
...
-		<a href="[[target]]" on-mouseover="showHoverMenu" on-mouseout="hideHoverMenu">
+		<d2l-ajax
+			id="update-preferences"
+			url="[[_updatePreferencesUrl]]"
+			method="[[_updatePreferencesMethod]]"
+			body="[[_updatePreferencesBody]]"
+			headers='{ "Accept": "application/vnd.siren+json" }'
+			on-response="updatePreferencesOnResponse"></d2l-ajax>
+
+		<a href="" on-mouseover="hoverCourseTile" on-mouseout="unhoverCourseTile">
```
Comment (d2l-joelp) [inline d2l-course-tile.html]: Does the course tile not navigate anymore? Curious about `href=""`.

File: d2l-course-tile.html [hunk 4]
```diff
@@ -1,86 +1,143 @@
...
+			},
+			_enrollmentEntityChanged: function (entity) {
+				if (entity.hasLink('course-image')) {
+					var courseImageLink = entity.getLinkByRel('course-image');
+					var courseImage = this.$$('.course-image img');
+					Polymer.dom(courseImage).setAttribute('src', courseImageLink.href);
 				}
+
+				var anchor = this.$$('a');
+				Polymer.dom(anchor).setAttribute('href', '/d2l/home/' + entity.properties.id);
```
Comment (d2l-joelp) [inline d2l-course-tile.html]: Oh, I see. nvm my comment above. But it feels like this can still be done with binding, no?

File: d2l-course-tile.html [hunk 5]
```diff
@@ -1,86 +1,143 @@
...
 			},
 			ready: function () {
 				this.setAttribute('role', 'link');
 			},
-			showHoverMenu: function() {
+			updatePreferencesOnResponse: function (response) {
+				if (response.detail.status === 200) {
+					var parser = document.createElement('d2l-siren-parser');
+					var preferencesEntity = parser.parse(response.detail.xhr.response);
+					this._pinned = preferencesEntity.properties.hasOwnProperty('pinDate');
```
Comment (d2l-joelp) [inline d2l-course-tile.html]: Do we really want to wait until the call comes back to update the pin state? To make the UI feel more responsive I feel like we should assume it worked but if it fails we can revert the change to pin state. Thoughts?

File: d2l-all-courses.html (hunk 1 above)
Comment (ryantmer) [inline d2l-all-courses.html]: I did, although I cleaned up the order here a bit, so it lost it it seems. I shall remove!

File: d2l-course-tile.html (hunk 4 above)
Comment (ryantmer) [inline d2l-course-tile.html]: Yeah, definitely a good case for binding I'd say. I only included the `href=""` since strictly speaking an anchor requires an `href`, pretty sure. But yeah, binding is mo' betta.

File: d2l-course-tile.html (hunk 5 above)
Comment (ryantmer) [inline d2l-course-tile.html]: In the `pinClickHandler` I change the value of `this._pinned` immediately, which calls `_pinnedChanged` and does update the UI immediately. In theory, this line here should have no effect, since `this._pinned` should already be set to the correct value.

File: d2l-course-tile.html [hunk 6]
```diff
@@ -1,86 +1,143 @@
...
+				}
+			},
+			pinClickHandler: function(e) {
+				// Prevent the click from triggering navigation to the course
+				e.preventDefault();
+
+				this._pinned = !this._pinned;
+
+				var updatePreferences = this.$$('#update-preferences');
```
Comment (ryantmer) [inline d2l-course-tile.html]: Aw. Yeah probably. Forgot we can't really use the advantages of Shadow DOM just yet :<

File: d2l-course-tile.html (hunk 5 above)
Comment (d2l-joelp) [inline d2l-course-tile.html]: My bad. Missed that.

Comment (ryantmer) [inline d2l-course-tile.html]: That said, I realize now that if the value comes back as a non-200, I'm not setting the icon to go back to the previous one. I'll add that in.

File: d2l-course-tile.html [hunk 7]
```diff
@@ -1,86 +1,143 @@
...
+
+				var updatePreferencesBody = {};
+				this._updatePreferencesFields.forEach(function(field) {
```
Comment (d2l-joelp) [inline d2l-course-tile.html]: For the future, this type of stuff might be useful to put in `d2l-siren-parser`. And by 'this type of stuff' I mean things to help you build a correct request when executing actions.

Comment (ryantmer) [inline d2l-course-tile.html]: Yeah, this came up a while ago as well. I think it would be handy to have a module that does it, but I think it would be good to have it separate from the parser itself, since it's pretty simple and network requests are less so. But something that complements it, fo sho.

Comment (d2l-joelp) [inline d2l-course-tile.html]: Sorry, I just meant to help create the body, not actually send the request. Seems like every time you'd want to call an action you'll loop through the fields to add them all to the body and set the appropriate default values. How you end up calling the action is app specific for sure.

Comment (ryantmer) [inline d2l-course-tile.html]: Ah! I gotcha. Although much simpler, I still feel that that would be more at home in a `node-siren-request` or something, as the parser is more for deconstructing rather than constructing.

Comment (d2l-joelp) [inline d2l-course-tile.html]: Good point. I'm with ya.

**Comment by ryantmer** [comment]
Unfortunately the pin/unpin icon still has somewhat wonky spacing, but this is caused by the SVG itself rather than any sort of padding/margin, so I'll have to check with @thehappypixel on that once he gets back.
